File: src/com/leozhi/topic125/Solution04.java
package com.leozhi.topic125;

/**
 * @author leozhi
 * 左右指针
 * 通过
 * 5ms
 */
public class Solution04 {
    public boolean isPalindrome(String s) {
        int n = s.length();
        int left = 0, right = n -1;
        while (left < right) {
            while (left < right && !Character.isLetterOrDigit(s.charAt(left))) {
                left++;
            }
            while (left < right && !Character.isLetterOrDigit(s.charAt(right))) {
                right--;
            }
            if (left < right) {
                if (Character.toLowerCase(s.charAt(left)) != Character.toLowerCase(s.charAt(right))) {
                    return false;
                }
                left++;
                right--;
            }
        }
        return true;
    }
}

File: src/com/leozhi/topic852/Solution04.kt
package com.leozhi.topic852

/**
 * @author leozhi
 * 二分查找
 * 通过
 * 288ms
 */
class Solution04 {
    fun peakIndexInMountainArray(arr: IntArray): Int {
        var left = 0
        var right = arr.size - 1
        while (left < right) {
            val mid = left + (right - left) / 2
            if (arr[mid] < arr[mid + 1]) {
                left = mid + 1
            } else {
                right = mid
            }
        }
        return left
    }
}